### ISSUE TITLE
TRM’s Overfitting Behavior and Model Scaling

### ISSUE THREAD

**Issue by gaoxin492** [SamsungSAILMontreal/TinyRecursiveModels#19]
Based on the training observations, I have a few questions:

The performance gap between training and test sets is quite large — the **training exact accuracy is close to 1.0, while the validation accuracies are around (~20%, ~1%, ~80%, ~60%) for the four datasets respectively.**
Would this be considered a form of **overfitting**? Each task already requires non-trivial compute resources, so it seems the model is fitting the training data almost perfectly but not generalizing well.

<img width="300" height="200" alt="Image" src="https://github.com/user-attachments/assets/42afa6cb-5004-4cb2-9f28-ccc30de8f21c" />

<img width="300" height="200" alt="Image" src="https://github.com/user-attachments/assets/ff366465-4a92-47c9-aecc-4daa1b9e75cb" />

Also, since the paper mentions that “two latent variables” and “two layers” work best, I’m curious how this model could **scale** further.
From Table 3, it seems that increasing the number of recursive steps or depth yields diminishing returns. Does that mean TRM’s performance is inherently bounded by its recursive structure? How could we scale this model effectively — by using a much larger parameter count (e.g., >1B), a wider architecture, or perhaps a larger and more diverse training dataset?

**Comment by asyua-ye** [comment]
> Based on the training observations, I have a few questions:
> 
> The performance gap between training and test sets is quite large — the **training exact accuracy is close to 1.0, while the validation accuracies are around (~20%, ~1%, ~80%, ~60%) for the four datasets respectively.** Would this be considered a form of **overfitting**? Each task already requires non-trivial compute resources, so it seems the model is fitting the training data almost perfectly but not generalizing well.
> 
> <img alt="Image" width="300" height="200" src="https://private-user-images.githubusercontent.com/74125950/501474633-42afa6cb-5004-4cb2-9f28-ccc30de8f21c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEwMjU3MTksIm5iZiI6MTc2MTAyNTQxOSwicGF0aCI6Ii83NDEyNTk1MC81MDE0NzQ2MzMtNDJhZmE2Y2ItNTAwNC00Y2IyLTlmMjgtY2NjMzBkZThmMjFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIxVDA1NDMzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyYWQ3NTcyNWI3NjE1Y2ZkOGU3Y2NiYTU5MzBlOTgxOTdiNDdjMjdkZjY3NWQzYzUwMGYzMTMxYjcxNGEzOWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-KxQ1dPQSjQGoLBddGD7WkV21Gyr3X1JqbP4SS5BvHk"> <img alt="Image" width="300" height="200" src="https://private-user-images.githubusercontent.com/74125950/501474737-ff366465-4a92-47c9-aecc-4daa1b9e75cb.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEwMjU3MTksIm5iZiI6MTc2MTAyNTQxOSwicGF0aCI6Ii83NDEyNTk1MC81MDE0NzQ3MzctZmYzNjY0NjUtNGE5Mi00N2M5LWFlY2MtNGRhYTFiOWU3NWNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIxVDA1NDMzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyZDk3M2FlMTY1OTFmMjIzZDI3Y2Y1ZjZlYWYwZDM5M2QyZmE4MGMzNDdmOTM4MTlhZDZhOTAwN2Q3ZDEwOTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vhTQjJb2Kj3Ew48D7g2xTyp1wB6kORINZxzJk4W1DKQ">
> Also, since the paper mentions that “two latent variables” and “two layers” work best, I’m curious how this model could **scale** further. From Table 3, it seems that increasing the number of recursive steps or depth yields diminishing returns. Does that mean TRM’s performance is inherently bounded by its recursive structure? How could we scale this model effectively — by using a much larger parameter count (e.g., >1B), a wider architecture, or perhaps a larger and more diverse training dataset?


I’d like to ask: which specific tasks do the validation accuracies of approximately (~20%, ~1%, ~80%, ~60%) correspond to? Assuming that the ~20% and ~1% figures come from ARC tasks, does that imply the model actually performs even worse on the test set? And does this suggest that the model’s generalization capability is highly dependent on the nature of the individual tasks?

**Comment by Chase08765** [comment]
What’s your highest accuracy on ARC-2 when using TRM?
If you achieve any accuracy, please leave a message here.

**Comment by olivkoch** [comment]
I am able to reproduce the results of the paper.

I get 85% exact accuracy on Sudoku Extreme and 75% exact accuracy on Maze Hard (on val sets).

I did not try to reproduce results on ARC-AGI as the training cost is significant.

I get these results both with this repo and with my own implementation of TRM.

https://github.com/olivkoch/nano-trm
